libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/StackLimitEvader.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+// Walks the call graph using a Depth-First-Search assigning memory offsets to variables.
+// - The leaves of the call graph will get the lowest offsets, increasing towards the root.
```

Comment:
Would it be very difficult to use small offsets at the trunk instead? This seems more intuitive to me.

Reply:
They have to be assigned in order from leaves to root, everything else would vastly complicate the algorithm I think. But I could of course just reverse the order afterwards, i.e. just use assigned slots as "end of reserved memory minus slot" instead or something like that... but that would probably make things unnecessarily more complicated in the end, wouldn't it :-)?

I can't do it the other way round in the algorithm itself, since leaves have to be assigned first and I don't yet know how much space I'll need in total, when assigning the leaves.

Reply:
The only place where this is relevant is StackToMemoryMover::memoryOffset isn't it? All the other places just count number of slots or use "slot numbers". How these slots are then assigned to actual memory offsets is irrelevant, as long as they are unique.

Reply:
Yes, sure - on the other hand it's quite natural to map the "slot numbers" just linearly to memory offsets... but if you think it's nicer and not confusing, then I can calculate the actual offsets in reverse order from the slots, I don't really mind.

Reply:
If you look at optimized code, I think it would be less confusing to see small memory offsets at the beginning.

Reply:
Did it in a separate commit.